### PR TITLE
Structure Change - Split functions into their own files

### LINKED ISSUE #17: Structual Change: Split each individual function into its own file
I have been contemplating support for pick and choose import support. To achieve this we first need to split each function into its own file. That also means taking the type definitions into said file.

Our index file then imports all other files.

Example structure
```
src/
 - index.ts
 - sum.ts
 - avg.ts
```
Example file (sum.ts)

```typescript

const sum: <T>(callbackfn?: (value: T) => number, thisArg?: any[]) => number = function (
  func = x => (x as unknown) as number,
  thisArg = this
) {
  return thisArg.reduce<number>((acc, t: any) => acc + func(t), 0);
};

/**
 * Shorthand for creating a summation by a single attribute
 *
 * Examples:
 * ```typescript
 * const totalScore = myArr.sum(x => x.score);
 * const totalScore = [1,2,3].sum();
 * ```
 */
Array.prototype.sum = sum;

```

Example index.ts
```typescript
import "./sum";
import "./avg";
```

### PR DESCRIPTION
**Pull request by JeppeTrip**
Fixes #17 .
Moved the array extensions into their individual files.

### CODE REVIEW

**Comment by objmj-itminds** [comment]
The types.ts file should after a successful migration can be deleted.

Also. Can the utils files such as be moved into a utils dir?

**Comment by JeppeTrip** [comment]
> Also. Can the utils files such as be moved into a utils dir?

What utils files should be moved into the utils directory ?

**Comment by objmj-itminds** [comment]
> What utils files should be moved into the utils directory ?
customErrors.ts 
typeGuards.ts

**Review by objmj-itminds — CHANGES_REQUESTED** [review]
There is a difference in the way the typing is handled per file.
Some use 
```typescript
declare global { 
...
}
``` 

Other do not.
If this is on purpose please leave a comment in the code explaining the use.

**Comment by JeppeTrip** [comment]
> 
> 
> There is a difference in the way the typing is handled per file.
> Some use
> 
> ```ts
> declare global { 
> ...
> }
> ```
> 
> Other do not.
> If this is on purpose please leave a comment in the code explaining the use.

I have added an explanation to some of the relevant files. 
The reason is basically that TS defines a module as any file with and import or export statement. TS furthermore scopes their modules and for us to modify a type (such as Array) existing in the global scope we have to declare our module as being in the global scope. 
Files without import or export statements are treated as scripts in the global scope and it is therefor not required (and not possible due to TS) to declare it being in the global scope, since it already exists there.
https://www.typescriptlang.org/docs/handbook/modules.html
https://www.typescriptlang.org/docs/handbook/declaration-files/templates/global-modifying-module-d-ts.html

**Comment by objmj-itminds** [comment]
We have discussed using directories for the move
```
/
- src/
- - sum/
- - - index.ts
- - - README.md
- - average/
- - - index.ts
- - - README.md
```

File: tsconfig.json [hunk 1]
```diff
@@ -10,6 +10,6 @@
     "outDir": "build",
     "resolveJsonModule": true
   },
-  "include": ["src/**/*.ts"],
+  "include": ["src/**/*.ts", "src/utils/typeGuards.ts", "src/utils/customErrors.ts"],
```
Comment (objmj-itminds) [inline tsconfig.json:13]: The added cases are covered by the first statement. Might be an editor suggestion.

**Review by objmj-itminds — APPROVED** [review]
I will create some other chores to add details in each README but I like this. 
I think we need to prio this PR so the others can get to target the changes made from this.

**Pull request merged**